dataset/test/data_array_test.cpp

```diff
@@ -137,3 +137,30 @@ TEST_F(DataArrayTest, full_slice) {
   EXPECT_TRUE(slice.masks()["mask"].is_same(a.masks()["mask"]));
   EXPECT_TRUE(slice.attrs()[Dim("attr")].is_same(a.attrs()[Dim("attr")]));
 }
+
+TEST_F(DataArrayTest, self_nesting) {
+  DataArray inner{makeVariable<double>(Dims{Dim::X}, Shape{2}, Values{1, 2})};
+  Variable var = makeVariable<DataArray>(Values{inner});
+
+  DataArray nested_in_data{var};
+  ASSERT_THROW_DISCARD(var.value<DataArray>() = nested_in_data,
+                       std::invalid_argument);
+
+  DataArray nested_in_coord{
+      makeVariable<double>(Dims{Dim::X}, Shape{2}, Values{3, 4})};
+  nested_in_coord.coords().set(Dim::X, var);
+  ASSERT_THROW_DISCARD(var.value<DataArray>() = nested_in_coord,
+                       std::invalid_argument);
```

Comment:
So if I understand correctly this and the ones below detects cases such as

```python
da.coords['abc'].value = da
```

What about setting up the same, but in reverse order:

```python
da.coords['abc'] = sc.scalar(da)
```

which I believe is much more likely to occur in practice.

Reply:
This is fine. `sc.scalar(da)` makes a shallow copy of `da`. So the new coord does not have 'abc' as a coord itself.

Reply:
Is it fine though? What happens now if you `del da`? Doesn't it leak?

Reply:
Why would it?
I can this through valgrind with and without `del` and it shows no difference. Also, there is no difference whether I add the coord or not. Which makes sense, I think,

Reply:
Right, I think you'd need a more contrieved example.